testsuite/forge/src/runner.rs

```diff
@@ -0,0 +1,262 @@
+        summary.write_starting_msg()?;
+
+        let mut rng = ::rand::rngs::StdRng::from_seed([0; 32]);
+        let mut swarm = self.factory.launch_swarm();
+
+        // Run PublicUsageTests
+        for test in self.tests.public_usage_tests {
+            let mut public_ctx =
+                PublicUsageContext::new(CoreContext::from_rng(&mut rng), swarm.public_info());
+            let result = run_test(|| test.run(&mut public_ctx));
+            summary.handle_result(test.name().to_owned(), result)?;
+        }
+
+        // Run AdminTests
+        for test in self.tests.admin_tests {
+            let mut admin_ctx =
+                AdminContext::new(CoreContext::from_rng(&mut rng), swarm.admin_info());
+            let result = run_test(|| test.run(&mut admin_ctx));
+            summary.handle_result(test.name().to_owned(), result)?;
+        }
+
+        summary.write_summary()?;
+
+        if !summary.success() {
+            process::exit(101);
+        }
+
+        Ok(())
+    }
+}
+
+enum TestResult {
+    Ok,
+    Failed,
+    FailedWithMsg(String),
+}
+
+fn run_test<F: FnOnce() -> Result<()>>(f: F) -> TestResult {
+    match ::std::panic::catch_unwind(::std::panic::AssertUnwindSafe(f)) {
+        Ok(Ok(())) => TestResult::Ok,
+        Ok(Err(e)) => TestResult::FailedWithMsg(format!("{:?}", e)),
+        Err(_) => TestResult::Failed,
+    }
+}
+
+struct TestSummary {
+    stdout: StandardStream,
+    total: usize,
+    filtered_out: usize,
+    passed: usize,
+    failed: Vec<String>,
+}
+
+impl TestSummary {
+    fn new(total: usize, filtered_out: usize) -> Self {
+        Self {
+            stdout: StandardStream::stdout(ColorChoice::Auto),
+            total,
+            filtered_out,
+            passed: 0,
+            failed: Vec::new(),
+        }
+    }
+
+    fn handle_result(&mut self, name: String, result: TestResult) -> io::Result<()> {
+        write!(self.stdout, "test {} ... ", name)?;
+        match result {
+            TestResult::Ok => {
+                self.passed += 1;
+                self.write_ok()?;
+            }
+            TestResult::Failed => {
+                self.failed.push(name);
+                self.write_failed()?;
+            }
+            TestResult::FailedWithMsg(msg) => {
+                self.failed.push(name);
+                self.write_failed()?;
+                writeln!(self.stdout)?;
+
+                write!(self.stdout, "Error: {}", msg)?;
+            }
+        }
+        writeln!(self.stdout)?;
+        Ok(())
+    }
+
+    fn write_ok(&mut self) -> io::Result<()> {
+        self.stdout
+            .set_color(ColorSpec::new().set_fg(Some(Color::Green)))?;
+        write!(self.stdout, "ok")?;
+        self.stdout.reset()?;
+        Ok(())
+    }
+
+    fn write_failed(&mut self) -> io::Result<()> {
+        self.stdout
+            .set_color(ColorSpec::new().set_fg(Some(Color::Red)))?;
+        write!(self.stdout, "FAILED")?;
```

Comment:
Just a minor nit: `Failed` and `Success`?  And maybe the same case between both?

Reply:
This bit is actually to mimic what the rust test harness does

Reply:
Hmm, a strange choice by Rust, but that's fine